src/Expander.hs

```diff
@@ -340,39 +343,41 @@ visibleBindings = do
 
 
 
-allMatchingBindings :: Text -> ScopeSet -> Expand [(ScopeSet, Binding)]
+allMatchingBindings :: Text -> ScopeSet -> Expand (Seq (ScopeSet, Binding))
 allMatchingBindings x scs = do
-  namesMatch <- view (at x . non []) <$> visibleBindings
+  namesMatch <- view (at x . non mempty) <$> visibleBindings
   p <- currentPhase
-  let scopesMatch =
-        [ (scopes, b)
-        | (scopes, b, _) <- namesMatch
-        , ScopeSet.isSubsetOf p scopes scs
-        ]
+  let scopesMatch = fmap (\(s,b,_) -> (s,b))
+        . Seq.filter (\(scopes, _, _) -> ScopeSet.isSubsetOf p scopes scs)
+        $ namesMatch
   return scopesMatch
 
 
 
-checkUnambiguous :: ScopeSet -> [ScopeSet] -> Ident -> Expand ()
+checkUnambiguous :: ScopeSet -> (Seq ScopeSet) -> Ident -> Expand ()
 checkUnambiguous best candidates blame =
   do p <- currentPhase
      let bestSize = ScopeSet.size p best
-     let candidateSizes = map (ScopeSet.size p) (nub candidates)
-     if length (filter (== bestSize) candidateSizes) > 1
-       then throwError (Ambiguous p blame candidates)
-       else return ()
+     let candidateSizes = Seq.fromList $ fmap (ScopeSet.size p) (nub $ toList candidates)
+     when (Seq.length (Seq.filter (== bestSize) candidateSizes) > 1)
+       $ throwError (Ambiguous p blame candidates)
 
 resolve :: Ident -> Expand Binding
 resolve stx@(Stx scs srcLoc x) = do
   p <- currentPhase
   bs <- allMatchingBindings x scs
   case bs of
-    [] ->
+    Seq.Empty ->
       throwError (Unknown (Stx scs srcLoc x))
     candidates ->
-      let best = maximumOn (ScopeSet.size p . fst) candidates
-      in checkUnambiguous (fst best) (map fst candidates) stx *>
-         return (snd best)
+      let
+        check = ScopeSet.size p . fst
+        c_last :: Seq a -> a
+        c_last (_xs Seq.:|> lst) = lst
+        c_last _                 = error "resolve: this is impossible"
+        (best_scpset, best_bind) = c_last $ Seq.sortOn check candidates
```

Comment:
`minimum` being asymptotically faster than `fst . sort` being the prototypical example of the benefit of laziness, I'm surprised to see that this time it's the opposite, `maximum` is slower than `last . sort`! Can you explain why we don't get the expected asymptotic gain from laziness in this case?

Reply:
No I think it is just slower but non-dominating. But a Data.Sequence has log time access because its a finger tree, so its not as bad as a linked list.

Reply:
see 2e45103d571a88c9ceaf247a8b715d751c0f3f86